File: maze.cpp
// maze.cpp
// Implementations for Maze class
// Author: Juan Arias
//
// The Maze class represents a 2D maze read from a text file which specifies
// width, height, exit row, exit location and the characters representing the
// maze. The Maze is capable of determinning whether a given location
// is a space, wall, visited or path and
// capable of marking a given location as visited or part of the path.

#include "maze.h"
#include <fstream>
#include <string>
#include <iomanip>

// Friend function
// Overloaded output operator
// Prints the maze in the format specified by the its file
std::ostream& operator<<(std::ostream& out, const Maze& maze) {
	
	out << std::left << std::setw(5) << maze.width << maze.height
		<< std::endl
		<< std::left << std::setw(5) << maze.exitRow << maze.exitCol
		<< std::endl;

	for (int i(0); i < maze.height; ++i) {

		for (int j(0); j < maze.width; ++j) {
		
			out << maze.array[i][j];
		
		}

		out << std::endl;

	}
	
	return out;
}

// Maze constructor requiring a valid file name as parameter
// Initalizes fields then calls helper method initializeArray
Maze::Maze(const std::string& fileName) :array(nullptr) {

	std::ifstream file(fileName);

	if (!file.is_open()) {

		std::cout << "Error in filename" << std::endl;

		throw 99;
	}

	file >> this->width;
	file >> this->height;
	file >> this->exitRow;
	file >> this->exitCol;
	file.ignore();

	this->initializeArray(file);

	file.close();
}

// Copy constructor
// Initializes array to nullptr,
// uses overloaded assignment operator to
// copy state of parameter other
Maze::Maze(const Maze& other) :array(nullptr) {

	*this = other;
}

// Overloaded assignment operator
// Copies the state of parameter other,
// checks for self assignment and uninitialized pointer,
// returns *this by reference
Maze& Maze::operator=(const Maze& other) {

	if (this != &other) {

		if (this->array != nullptr) {

			this->clear();
		}

		this->width   = other.width;
		this->height  = other.height;
		this->exitRow = other.exitRow;
		this->exitCol = other.exitCol;

		this->array = new LOCATION*[height];

		for (int i(0); i < this->height; ++i) {

			this->array[i] = new LOCATION[this->width];

			for (int j(0); j < this->width; ++j) {

				this->array[i][j] = other.array[i][j];
			}
		}
	}

	return *this;
}

// Destroys Maze
// Deallocates all dynamic memory
Maze::~Maze() {

	this->clear();
}

// Returns true if the given position in the Maze is clear (space),
// otherwise false
bool Maze::IsClear(int row, int col) const {
	
	return (this->inBounds(row, col)) && (this->array[row][col] == CLEAR);

}

// Returns true if the given position in the Maze is wall ('x'),
// otherwise false 
bool Maze::IsWall(int row, int col) const {

	return (this->inBounds(row, col)) && (this->array[row][col] == WALL);

}

// Returns true if the given position in the Maze is path ('*'),
// otherwise false 
bool Maze::IsPath(int row, int col) const {

	return (this->inBounds(row, col)) && (array[row][col] == PATH);

}

// Returns true if the given position in the Maze is visited ('+'),
// otherwise false 
bool Maze::IsVisited(int row, int col) const {

	return (this->inBounds(row, col)) && (array[row][col] == VISITED);

}

// Returns true if given position is exit of Maze,
// otherwise false
bool Maze::IsExit(int row, int col) const {

	return (row == this->exitRow) && (col == this->exitCol);

}

// Marks the Maze at the given position with *
void Maze::MarkAsPath(int row, int col) {

	this->array[row][col] = PATH;
}

// Mark the Maze at the given position with +
void Maze::MarkAsVisited(int row, int col) {

	this->array[row][col] = VISITED;
}

// Initalizes 2D dynamic array with
// parameter file as the Maze's text file stream
void Maze::initializeArray(std::ifstream& file) {

	std::string line;

	this->array = new LOCATION*[height];

	for (int i(0); i < this->height; ++i) {

		std::getline(file, line);

		this->array[i] = new LOCATION[width];
		
		for (int j(0); j < this->width; ++j) {

			this->array[i][j] = static_cast<LOCATION>(line[j]);
		}
	}
}

// Returns true if position is in bounds of Maze,
// otherwise false
bool Maze::inBounds(int row, int col) const {

	return (0 <= row) && (row <= this->height) && (0 <= col) &&
		   (col <= this->width);

}

// Deallocates dynamic memory
void Maze::clear() const {

	for (int i(0); i < this->height; ++i) {

		delete[] this->array[i];
	}

	delete[] this->array;
}

File: Maze.java
/**
 * Maze.java
 * 
 * @author Wizzol
 * 
 * Maze shits
 */
package css342;

import java.util.*;
import java.util.zip.DataFormatException;
import java.io.*;

/**
 * Maze class
 * 
 * @author Juan
 * 
 * Represents a virtual maze created from a text file
 */
public class Maze {
	
	// 2D array to store maze chars
	private MAZETXT[][] matrix;
	
	// Exit location
	private int exitRow, exitCol;
	
	/**
	 * Enum to control type safety of input chars
	 * @author Juan
	 */
	private enum MAZETXT { 
		
		// Only chars allowed in maze text file
		SPACE(' '), WALL('x'), VISITED('+'), PATH('*');
		
		// Char variable
		private char c;
		
		/**
		 * Constructs MAZETXT constant with char input
		 * @param c The char input
		 */
		private MAZETXT(char c) { this.c = c; }
		
		/**
		 * Turns MAZETXT into char
		 * @return char of MAZETXT constant
		 */
		public char toChar() { return this.c; }
		
		/**
		 * Static function to get the right MAZETXT constant from a char
		 * @param c The char input
		 * @return corresponding MAZETXT constant
		 */
		protected static MAZETXT fromChar(char c) {
			
			switch (c) {
			
			case ' ':
				
				return SPACE;
				
			case 'x':
			
				return WALL;
			
			case '+':
				
				return VISITED;
				
			case '*':
				
				return PATH;
				
			default:
				
				return null;
			}
		}
	}
	
	/**
	 * Constructs Maze with given fileName
	 * @param fileName The name of the maze file
	 * @throws FileNotFoundException
	 * @throws DataFormatException
	 */
	public Maze(String fileName) throws FileNotFoundException, DataFormatException {
		
		Scanner fileScan = new Scanner(new File(fileName));
		
		if (!fileScan.hasNextInt()) {
			
			fileScan.close();
			
			throw new DataFormatException();
		}
		
		this.initializeFields(fileScan.nextInt(), fileScan.nextInt(),
							  fileScan.nextInt(), fileScan.nextInt());
		
		fileScan.nextLine();
		
		this.extractMaze(fileScan);
		
		fileScan.close();
	}
	
	/**
	 * Determines whether a spot in the Maze is clear for moving
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if clear, false if not
	 */
	public boolean isClear(int row, int col) {
		
		return this.inBounds(row, col) && this.matrix[row][col] == MAZETXT.SPACE;
	}
	
	/**
	 * Determines whether a spot in the Maze is a wall
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if wall, false if not
	 */
	public boolean isWall(int row, int col) {
		
		return this.inBounds(row, col) && this.matrix[row][col] == MAZETXT.WALL;
	}
	
	/**
	 * Determines whether a spot in the Maze is a part of the
	 * solution path
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if path, false if not
	 */
	public boolean isPath(int row, int col) {
		
		return this.inBounds(row, col) && this.matrix[row][col] == MAZETXT.PATH;
	}
	
	/**
	 * Determines whether a spot in the Maze has been
	 * already visited
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if visited, false if not
	 */
	public boolean isVisited(int row, int col) {
		
		return this.inBounds(row, col) && this.matrix[row][col] == MAZETXT.VISITED;
	}
	
	/**
	 * Determines whether a spot in the Maze is the exit
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if exit, false if not
	 */
	public boolean isExit(int row, int col) {
		
		return this.inBounds(row, col) && (row == this.exitRow) && (col == this.exitCol);
	}
	
	/**
	 * Determines whether a spot in the Maze is a part of the
	 * solution path
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if path, false if not
	 */
	public void markAsPath(int row, int col) {
		
		this.matrix[row][col] = MAZETXT.PATH;
	}
	
	/**
	 * Determines whether a spot in the Maze has been
	 * already visited
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if visited, false if not
	 */
	public void markAsVisited(int row, int col) {
		
		this.matrix[row][col] = MAZETXT.VISITED;
	}
	
	/**
	 * Converts maze to printable string
	 * @return String representation of maze with chars
	 */
	@Override
	public String toString() {
		
		StringBuffer s = new StringBuffer();
		
		for (MAZETXT[] array : this.matrix) {
			
			for (MAZETXT constant : array) {
				
				s.append(constant.toChar());
			}
			
			s.append('\n');
		}
		
		return s.toString();
	}
	
	/**
	 * Helper method for constructor
	 * @param width The width of the maze
	 * @param height The height of the maze
	 * @param exitRow The row number of the exit
	 * @param exitCol The column number of the exit
	 */
	private void initializeFields(int width, int height, int exitRow, int exitCol) {
		
		this.matrix = new MAZETXT[height][width];
		this.exitRow = exitRow;
		this.exitCol = exitCol;
	}
	
	/**
	 * Extracts chars from text file into this maze
	 * @param fileScan The file scanner
	 */
	private void extractMaze(Scanner fileScan) {
		
		for (int row = 0; row < this.matrix.length && fileScan.hasNextLine(); ++row) {
			
			String line = fileScan.nextLine();
			
			for (int col = 0; col < this.matrix[0].length; col++) {
				
				this.matrix[row][col] = MAZETXT.fromChar(line.charAt(col));
			}
		}
	}
	
	
	/**
	 * Helper for other boolean methods to check that
	 * given spot is within boundaries
	 * @param row The row of the spot
	 * @param col The column of the spot
	 * @return boolean: true if in bounds, false if not
	 */
	private boolean inBounds(int row, int col) {
		
		return (row >= 0) && (row < this.matrix.length) && (col >= 0) &&
			   (col < this.matrix[row].length);
	}
}
